keps/sig-storage/2324-volume-snapshot-namespace-transfer/README.md

```diff
@@ -0,0 +1,543 @@
+to check the status of the transfer.
+
+- Can a transfer be performed when the PVC is having an operation performed on it?
+
+This issue is less of a concern for `VolumeSnapshots`, as operations are typically not performed on the
+snapshot, but for PVCs it is possible the user could be resizing or snapshotting the PVC during
+the transfer request.
+
+The current approach of using a Finalizer and deleting the source PVC should address this concern.
+For instance, the external-resizer can resume the operation on a PVC after the transfer is completed.
+We should ensure that other sidecars can continue once the transfer is completed as well.
+
+- As long as we reference the existing VolumeSnapshotContent, there should be no issues with encrypted VolumeSnapshotContents. The secret reference is stored on the VolumeSnapshotContent and VolumeSnapshotClass, not the VolumeSnapshot itself. If we take a different approach, such as re-creating the VolumeSnapshotContent, then we will need to ensure these credentials are propagated to the new content.
+- The VolumeSnapshot must be `readyToUse` before the transfer begins.
+- Any owners of the destination namespace must accept the transfer. By using the StorageTransfer API, this ensures an owner in the destination namespace is aware of the transfer.
+
+## Design Details
+
+This continues the discussion from [API for PVC Namespace Transfer](https://github.com/kubernetes/enhancements/pull/2110/) 
+in regards to the StorageTransfer API.
+
+A `StorageTransferRequest` resource will be created in the source namespace:
+
+```yaml
+apiVersion: storage.k8s.io/v1alpha1
+kind: StorageTransferRequest
+metadata:
+  name: request-name
+  namespace: source-namespace
+spec:
+  source:
+    name: foo
+    kind: VolumeSnapshot
+  secrets:
+    - name: secret-name
+      namespace: secret-namespace
+      type: secret-type # such as `csi.storage.k8s.io/provisioner-secret-name`
+  acceptName: accept-name
+  token: xyji6OwynW # This will be populated automatically by the controller
+  targetName: bar
+```
+
+`StorageTransferAccept` resource will be created in the target namespace:
+
+```yaml
+apiVersion: storage.k8s.io/v1alpha1
+kind: StorageTransferAccept
+metadata:
+  name: accept-name
+  namespace: target-namespace
+spec:
+  sourceNamespace: source-namespace
+  requestToken: xyji6OwynW
+  requestName: request-name
+```
+
+ When matching `StorageTransferRequest` and `StorageTransferAccept` resources are detected, the 
+ transfer process is initiated. We have the following matching criteria for the `StorageTransferRequest` 
+ named `r` and the `StorageTransferAccept` `a`:
+ - `r.metadata.name == a.spec.requestName`
+ - `r.metadata.namespace == a.spec.sourceNamespace`
+ - `r.spec.acceptName == a.metadata.name`
+ - `r.spec.token == a.spec.requestToken`
+
+Once matching is complete, a controller will begin the transfer process. For PVCs, this performs the following:
```

Comment:
@mhenriks @huffmanca

What happens if
- two `StorageTransferRequest`s are created for the same resource and `StorageTransferAccept` for each `StorageTransferRequest` is created almost at the same time,
- two `StorageTransferAccept`s for the same `StorageTransferRequest` are created almost at the same time in different namespaces,
- two `StorageTransferRequest`s that have the same `source.kind` and `targetName` are created, and `StorageTransferAccept`s for both of them are created in the same namespace,
- a `StorageTransferRequest` is deleted before transfer completes,
- a `StorageTransferAccept` is deleted before transfer completes,
- a resource with the same criteria (namespace, name, and so on) is created after a transfer for the intended resource is completed (do users need to ensure that `StorageTransferRequest` and `StorageTransferAccept` are deleted after a successful transfer)?

Reply:
> two StorageTransferRequests are created for the same resource and StorageTransferAccept for each StorageTransferRequest is created almost at the same time,
two StorageTransferAccepts for the same StorageTransferRequest are created almost at the same time in different namespaces,
two StorageTransferRequests that have the same source.kind and targetName are created, and StorageTransferAccepts for both of them are created in the same namespace,

I think these are all basically the same issue.  The transfer controller will process the first match it encounters and ignore others.  An annotation on the source can probably help us with the ignore case.  For applications with possibility of contention like this would probably be wise to have a standard naming scheme for requests/accepts and handle collisions that way

Reply:
> a StorageTransferRequest is deleted before transfer completes,
a StorageTransferAccept is deleted before transfer completes,

I imagine a pending and running phase for a transfer.  During the pending phase (waiting for no pods using the source, etc) a transfer is easily aborted.  Once a transfer is running, things should happen pretty quickly and finalizars can make sure the process completes.  However, if an error is encountered in the running phase, the transfer can be aborted and the system remains in the current state (no rollback)

Reply:
> a resource with the same criteria (namespace, name, and so on) in StorageTransferRequest is created after a transfer for the intended resource is completed (do users need to ensure that StorageTransferRequest and StorageTransferAccept are deleted after a successful transfer)?

Requests/Accepts should be deleted by the user or perhaps only an admin so that proper auditing occurs

Reply:
@mhenriks

Thank you for sharing your ideas. Can this information be added to the KEP so that we can discuss more on further corner cases?

On changing the spec, below points would be worth considering.

- What information is needed in annotations to mark PVC as transferring? 
  (One annotation that has identities for both `StorageTransferRequest` and `StorageTransferAccept`, or two separate annotations for each of them?)
- Do we need to add finalizer to both `StorageTransferRequest` and `StorageTransferAccept` to have a chance to properly remove the annotations on a transferring PVC for cancelation?
- Should we consider adding `targetNamespace` to `StorageTransferRequest.Spec`?
  (Current spec seems to allow one StorageTransferRequest to have multiple target namespace candidates, as a result, above annotation logic is likely to become complex. If we change the spec this way, we can focus on one annotation that has identity for only `StorageTransferRequest`.)
- Do `StorageTransferRequest` and `StorageTransferAccept` need to have `Status` fields to manage a pending and running phase?
  (And possibly, should we consider adding completed phase to avoid running multiple times after a successful transfer?)
- Can we consider a way to rollback even after it becomes running phase?


Also, an example for the further corner cases is how to block/work with operations by other controllers, like provisioning, attaching, snapshotting, resizing, etc ... (Do we just make other controllers understand the annotations or consider adding status fields to PVC for this purpose? And if we choose the latter, should information in above annotations also be moved to status fields?)